{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"1wO_BfX_tBEfIadQW8XQDPPNgC-FHAHX_","timestamp":1765564085695},{"file_id":"https://github.com/kimjc95/computational-chemistry/blob/main/Boltz_on_Colab.ipynb","timestamp":1763165302877}],"machine_shape":"hm","gpuType":"L4","collapsed_sections":["zY3uNlU410Mi","_2ro0HncbFfL","mhGnG1GEglwB","EjpvgFJEPdlB","G6XxZePbbV93","0jgu3H061gPp"]},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"},"accelerator":"GPU"},"cells":[{"cell_type":"markdown","source":["# Workshop on Structure-Based Drug Design on a Laptop Spring 2026\n","\n","# 3.1 Structure Prediction\n","\n","In this session we will prepare the structure of the protein-ligand complexes using an AI co-folding model called [`Boltz`](https://github.com/jwohlwend/boltz)\n","\n","Below, you will enter the sequence of the [`HPK1`](https://www.rcsb.org/structure/7L25) protein from the Protein Data Bank and the SMILES strings of the ligands assigned to you to generate the 3D structures of the complexes.\n","\n","The notebook below is a slightly modified copy of the [`Boltz on Colab`](https://github.com/kimjc95/computational-chemistry) notebook by [`Joo-Chan Kim`](https://github.com/kimjc95).\n","\n","To run the notebook:\n","\n","1. Connect to a Google Colab Runtime by clicking on the down-arrow next to the `Connect` button on the upper left, select `Change runtime type`, and selecting a `L4` GPU and `High-RAM`. Then click `Connect`.\n","2. Execute the first cell to install the required software components.\n","3. Execute the following cell after entering the name for the job, such as `hpk1-lig1`)\n","4. At the bottom of the cell, enter the sequence of the protein from the FASTA file of the HPK1 protein and the SMILES of one of the ligands assigned to you in the previous session (`2.1 Retrieve Ligand Smiles from CHEMBL`). Click `Confirm` when done.\n","5. Scroll down to `Run Prediction` and execute the `Create YAML file from the input data` and `Run prediction using Boltz-2` cells.\n","6. Finally execute `Download results as a zip file` cell to download the results.\n","7. Repeat the above for each ligand SMILE string making sure to change the job name accordingly each time.\n","8. Disconnect the Runtime and scroll down to section `3.2` for the instructions to continue the structure preparation in Maestro.\n","\n","\n"],"metadata":{"id":"v4vaYAnJgpdz"}},{"cell_type":"markdown","source":["# Boltz on Colab\n","2025-06-13 Joo-Chan Kim at [MSBL](https://msbl.kaist.ac.kr), KAIST\n","\n","This is a Google Colaboratory Notebook for running Boltz-2 at ease (BSD-3 license).\n","\n","Please cite DOI:[10.5281/zenodo.14881401](https://doi.org/10.5281/zenodo.14881401) if you have used my code in your research.\n","\n","Please report to [my GitHub](https://github.com/kimjc95/computational-chemistry/issues) or email (kimjoochan@kaist.ac.kr) if you encounter a bug.\n","\n","Choose a runtime and run below cells one by one."],"metadata":{"id":"iYszONHiYU7y"}},{"cell_type":"code","source":["#@title Install dependencies\n","#@markdown Select your runtime type. \\\n","#@markdown GPU runtime with a high capacity RAM is strongly recommended.\n","\n","runtime = \"GPU(L4 or T4)\" #@param [\"TPU\", \"GPU(A100)\", \"GPU(L4 or T4)\", \"CPU\"]\n","\n","import os\n","import subprocess\n","\n","print('Installing dependencies... ', end='')\n","dependencies = \"torch torchvision torchaudio numpy hydra-core pytorch-lightning \"\n","dependencies += \"rdkit dm-tree requests pandas types-requests einops einx fairscale \"\n","dependencies += \"mashumaro modelcif wandb click pyyaml biopython scipy numba gemmi \"\n","dependencies += \"scikit-learn chembl_structure_pipeline \"\n","dependencies += \"cuequivariance_ops_cu12 cuequivariance_ops_torch_cu12 cuequivariance_torch\"\n","\n","if runtime == \"GPU(L4 or T4)\":\n","    precision = '32-true'\n","else:\n","    precision = 'bf16-true'\n","\n","subprocess.run(\"pip install ipywidgets torch torchvision torchaudio\", shell=True)\n","subprocess.run(\"git clone https://github.com/jwohlwend/boltz.git\", shell=True)\n","subprocess.run(f\"sed -i 's/bf16-mixed/{precision}/g' /content/boltz/src/boltz/main.py\", shell=True)\n","subprocess.run(f\"pip install {dependencies}\", shell=True)\n","subprocess.run(\"cd boltz; pip install --no-deps -e .\", shell=True)\n","\n","print('done.')"],"metadata":{"id":"6j5dMx1VSNED","cellView":"form","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1765836834866,"user_tz":300,"elapsed":39068,"user":{"displayName":"Joe Z Wu","userId":"04248511503294646969"}},"outputId":"14da6975-0597-4649-d63b-6116dde3ba15"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Installing dependencies... done.\n"]}]},{"cell_type":"markdown","source":["# Input data"],"metadata":{"id":"zY3uNlU410Mi"}},{"cell_type":"code","source":["#@title Enter sequence input\n","#@markdown Type the job title name without blanks in the box below.\n","job_title = \"hpk1-lig1\" #@param {type:\"string\"}\n","#@markdown Run this cell and by using the interactive widgets below, enter the molecule sequence data.\n","\n","#@markdown For small molecule ligands or modified residues, you can enter the CCD ID (Chemical Compoenent Dictionary code) which can be looked upon the [PDBeChem website](https://www.ebi.ac.uk/pdbe-srv/pdbechem/).\n","\n","import ipywidgets as widgets\n","from IPython.display import display, HTML\n","import os\n","import re\n","import requests\n","from rdkit import Chem, RDLogger\n","from rdkit.Chem import Draw, AllChem\n","from google.colab import files\n","from Bio.PDB import MMCIFParser\n","from Bio.PDB.Polypeptide import is_aa\n","\n","\n","def validate_input(text, input_type)->bool:\n","    \"\"\"\n","    Validate the input text about the molecule info based on the specified input type.\n","    \"\"\"\n","    if input_type == 'protein':\n","        # Use RegEx to check all letters are in 20 canonical amino acid types\n","        return re.match(r'^[AC-IK-NP-TVWY]+$', text.upper()) is not None\n","    elif input_type == 'dna':\n","        # Use RegEx to check all letters are either A, C, G, or T\n","        return re.match(r'^[ACGT]+$', text.upper()) is not None\n","    elif input_type == 'rna':\n","        # Use RegEx to check all letters are either A, C, G, or U\n","        return re.match(r'^[ACGU]+$', text.upper()) is not None\n","    elif input_type == 'smiles':\n","        # Use RDKit to validate the SMILES string\n","        RDLogger.DisableLog('rdApp.*')\n","        try:\n","            mol = Chem.MolFromSmiles(text, sanitize=True)\n","        except:\n","            return False\n","        if mol is None:\n","            return False\n","        else:\n","            return True\n","    elif input_type == 'ccd':\n","        # Call header from PDBe website and check HTTP response status\n","        url = f\"https://files.rcsb.org/ligands/download/{text.upper()}_ideal.cif\"\n","        try:\n","            response = requests.head(url, timeout=5)\n","            return response.status_code == 200\n","        except requests.exceptions.RequestException:\n","            return False\n","\n","\n","def get_residue_ccd(polymer_type, sequence, index)->str:\n","    \"\"\"\n","    From polymer sequence and index, return the residue's CCD code.\n","    \"\"\"\n","    AA_codes = {\n","        'A': 'ALA', 'R': 'ARG', 'N': 'ASN', 'D': 'ASP', 'C': 'CYS',\n","        'E': 'GLU', 'Q': 'GLN', 'G': 'GLY', 'H': 'HIS', 'I': 'ILE',\n","        'L': 'LEU', 'K': 'LYS', 'M': 'MET', 'F': 'PHE', 'P': 'PRO',\n","        'S': 'SER', 'T': 'THR', 'W': 'TRP', 'Y': 'TYR', 'V': 'VAL'}\n","\n","    DNA_codes = {'A': 'DA', 'T': 'DT', 'G': 'DG', 'C': 'DC'}\n","\n","    if not index:\n","        return ''\n","\n","    position = int(index)\n","    if position < 1 or position > len(sequence):\n","        return ''\n","\n","    residue = sequence[position-1].upper()\n","\n","    if polymer_type == 'protein':\n","        return AA_codes[residue]\n","    elif polymer_type == 'dna':\n","        return DNA_codes[residue]\n","    else:\n","        return residue\n","\n","\n","def get_atom_names(ccd_id)->list:\n","    \"\"\"\n","    Return a list of official atom names from CCD ID.\n","    \"\"\"\n","    try:\n","        url = f\"https://files.rcsb.org/ligands/download/{ccd_id.upper()}_ideal.cif\"\n","\n","        response = requests.get(url)\n","        response.raise_for_status()\n","\n","        atom_names = []\n","        lines = response.text.split('\\n')\n","\n","        in_atom_section = False\n","        for line in lines:\n","            if line.startswith('_chem_comp_atom'):\n","                in_atom_section = True\n","                continue\n","\n","            if in_atom_section:\n","                if line.startswith('#') or line.strip() == '':\n","                    break\n","                parts = line.split()\n","                if len(parts) > 0:\n","                    if not parts[1].startswith('H'):\n","                        atom_names.append(parts[1])\n","\n","        if not atom_names:\n","             print(f\"Could not retrieve atom names for {ccd_id}.\")\n","\n","        return atom_names\n","\n","    except:\n","        return []\n","\n","\n","seq_data = []\n","mod_data = []\n","bond_data = []\n","binder = '' # global variable to save binder chain ID\n","pocket_data = []\n","contact_data = []\n","template_data = []\n","lig_select = None # global variable to save ligand chain ID used for affinity calculation\n","\n","\n","def is_it_polymer(chain)->bool:\n","    \"\"\"\n","    Check the given sequence data is a polymer or not.\n","    \"\"\"\n","    for s in seq_data:\n","        if s['chain'] == chain:\n","            return s['type'] in ['protein', 'dna', 'rna']\n","    return False\n","\n","\n","def return_res_name(chain, index)->str:\n","    \"\"\"\n","    From seq_data and mod_data, return a list of atom names of the given residue.\n","    \"\"\"\n","    for m in mod_data:\n","        if m['chain'] == chain and m['index'] == index:\n","            return m['ccd']\n","\n","    for s in seq_data:\n","        if s['chain'] == chain:\n","            return get_residue_ccd(s['type'], s['sequence'], index)\n","\n","    return ''\n","\n","\n","def return_atom_name_list(chain, index)->list:\n","    \"\"\"\n","    From seq_data and mod_data, return a list of atom names of the given residue.\n","    \"\"\"\n","    return get_atom_names(return_res_name(chain, index))\n","\n","\n","def get_polypeptide_chain_ids(cif_file):\n","\n","    parser = MMCIFParser(QUIET=True)\n","    try:\n","        structure = parser.get_structure(\"protein_structure\", cif_file)\n","    except Exception as e:\n","        print(f\"Error while parsing cif files: {e}\")\n","        return []\n","\n","    polypeptide_chains = set()\n","\n","    for model in structure:\n","        for chain in model:\n","            for residue in chain:\n","                if is_aa(residue):\n","                    polypeptide_chains.add(chain.id)\n","                    break\n","\n","    return sorted(list(polypeptide_chains))\n","\n","\n","\n","class modify_entries():\n","    \"\"\"\n","    Main module to show interactive widgets\n","    \"\"\"\n","    def __init__(self, container):\n","        self.container = container\n","\n","\n","    def remove_seq_entry(self, b):\n","        \"\"\"\n","        Operates when '-' button is pressed\n","        \"\"\"\n","        for i in range(1, len(self.container.children)-1):\n","            # when the minus button is hit\n","            if self.container.children[i].children[0].children[-1] == b:\n","                newList = []\n","                # rename chain IDs in alphabetical order\n","                for j in range(i+1, len(self.container.children)-1):\n","                    show_chain = widgets.Label(value= 'chain '+str(chr(ord('A')+j-2)))\n","                    newline = widgets.HBox([show_chain]+list(self.container.children[j].children[0].children[1:]))\n","                    newEntry = widgets.VBox([newline]+list(self.container.children[j].children[1:]))\n","                    newList.append(newEntry)\n","\n","                self.container.children = list(self.container.children[:i]) + newList + [self.container.children[-1]]\n","                break\n","\n","\n","    def add_seq_entry(self, b):\n","        \"\"\"\n","        Operates when '+' button is pressed\n","        \"\"\"\n","\n","        # Chain ID\n","        show_chain = widgets.Label(value= 'chain '+str(chr(ord('A')+len(self.container.children)-2)))\n","\n","        # Molecule type\n","        select_type = widgets.Dropdown(\n","            options=['protein', 'dna', 'rna', 'smiles', 'ccd'],\n","            description=' is ',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='120px'))\n","\n","        # Molecule info string\n","        enter_sequence = widgets.Text(\n","            description=' described as :',\n","            placeholder='MAKEY... or CC1=CC=CC=C1 or ATP',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='600px'))\n","\n","        # check for cyclic polymers\n","        cyclic = widgets.Checkbox(\n","            value=False,\n","            description=' cyclic? ',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='100px'))\n","\n","        # minus button\n","        remove_btn = widgets.Button(description='-', layout=widgets.Layout(width='30px'))\n","        remove_btn.on_click(self.remove_seq_entry)\n","\n","        # Error message\n","        message = widgets.HTML(value='', layout=widgets.Layout(width='600px', padding='5px'))\n","\n","        def validate_string(change):\n","            \"\"\"\n","            Validate the molecule info entered by the user.\n","            If the input is incorrect, show the red error message.\n","            \"\"\"\n","\n","            if select_type.value in ['smiles', 'ccd'] and cyclic.value:\n","                message.value = f\"<span style='color: red;'>Only polymers (protein, DNA, and RNA) can be specified as cyclic!</span>\"\n","            elif validate_input(enter_sequence.value, select_type.value):\n","                message.value = \"\"\n","            elif select_type.value in ['protein', 'dna', 'rna']:\n","                message.value = f\"<span style='color: red;'>Enter the valid {select_type.value} sequence.</span>\"\n","            else:\n","                message.value = f\"<span style='color: red;'>Enter the valid {select_type.value} string.</span>\"\n","\n","        enter_sequence.observe(validate_string, names='value')\n","        select_type.observe(validate_string, names='value')\n","        cyclic.observe(validate_string, names='value')\n","\n","        line = widgets.HBox([show_chain, select_type, enter_sequence, cyclic, remove_btn])\n","        entry = widgets.VBox([line, message])\n","\n","        self.container.children = list(self.container.children[:-1]) + [entry, self.container.children[-1]]\n","\n","\n","    def update_seq_data(self, b):\n","        \"\"\"\n","        Operates when the confirm button is pressed\n","        \"\"\"\n","        seq_data.clear()\n","        for i in range(1, len(self.container.children)-1):\n","            entry = self.container.children[i]\n","            line = entry.children[0]\n","            message = entry.children[1]\n","            # Filter invalid lines\n","            if message.value != '':\n","                continue\n","            # Filter empty lines\n","            if line.children[2].value == '':\n","                continue\n","\n","            seq = {'chain': line.children[0].value[-1],\n","                   'type': line.children[1].value,\n","                   'sequence': line.children[2].value,\n","                   'cyclic': line.children[3].value,\n","                   'msa':''}\n","            seq_data.append(seq)\n","\n","\n","    def remove_an_entry(self, b):\n","        \"\"\"\n","        Operates when '-' button is pressed, this time on modification / constraint entries\n","        \"\"\"\n","        for i in range(1, len(self.container.children)-1):\n","            if self.container.children[i].children[0].children[-1] == b:\n","                self.container.children = list(self.container.children[:i]) + list(self.container.children[i+1:])\n","                break\n","\n","\n","    def add_mod_entry(self, b):\n","        \"\"\"\n","        Operates when '+' button is pressed, this time on modification / constraint entries\n","        \"\"\"\n","        polymers = []\n","        for s in seq_data:\n","            if s['type'] in ['protein', 'dna', 'rna']:\n","                polymers.append(s['chain'])\n","\n","        # select chain\n","        select_chain = widgets.Dropdown(\n","            options=polymers,\n","            description='chain ',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='90px'))\n","\n","        # enter 1-based residue index\n","        enter_index = widgets.Text(\n","            description='position: ',\n","            placeholder = '0 <',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='300px'))\n","\n","        # show the current residue name\n","        show_res = widgets.Label(value='change from: -', layout=widgets.Layout(width='150px'))\n","\n","        # read CCD code\n","        enter_ccd = widgets.Text(\n","            description='to:',\n","            placeholder='CCD code',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='300px'))\n","\n","        # Minus button\n","        remove_btn = widgets.Button(description='-', layout=widgets.Layout(width='30px'))\n","        remove_btn.on_click(self.remove_an_entry)\n","\n","        # Error message\n","        message = widgets.HTML(value='', layout=widgets.Layout(width='600px', padding='5px'))\n","\n","        def validate_string(change):\n","            \"\"\"\n","            Validate the input entered by the user.\n","            If the input is incorrect, show the red error message.\n","            \"\"\"\n","            error_type = ''\n","            for s in seq_data:\n","                if s['chain'] == select_chain.value:\n","                    res = get_residue_ccd(s['type'], s['sequence'], enter_index.value)\n","                    if res == '':\n","                        error_type = 'index'\n","                        show_res.value = 'change from: -'\n","                    else:\n","                        show_res.value = f\"change from: {res}\"\n","                    break\n","\n","            if not validate_input(enter_ccd.value, 'ccd'):\n","                error_type = 'CCD'\n","\n","            for m in mod_data:\n","                if m['chain'] == select_chain.value:\n","                    if m['index'] == int(enter_index.value):\n","                        error_type = \"duplicate\"\n","                        break\n","\n","            if error_type == '':\n","                message.value = \"\"\n","            elif error_type == \"duplicate\":\n","                message.value = f\"<span style='color: red;'>Duplicate residue index.</span>\"\n","            else:\n","                message.value = f\"<span style='color: red;'>Enter the valid {error_type}.</span>\"\n","\n","        select_chain.observe(validate_string, names='value')\n","        enter_index.observe(validate_string, names='value')\n","        enter_ccd.observe(validate_string, names='value')\n","\n","        line = widgets.HBox([select_chain, enter_index, show_res, enter_ccd, remove_btn])\n","        entry = widgets.VBox([line, message])\n","\n","        self.container.children = list(self.container.children[:-1]) + [entry, self.container.children[-1]]\n","\n","\n","    def update_mod_data(self, b):\n","        \"\"\"\n","        Operates when the confirm button is pressed, this time on modification / constraint entries\n","        \"\"\"\n","        mod_data.clear()\n","        for i in range(1, len(self.container.children)-1):\n","            entry = self.container.children[i]\n","            line = entry.children[0]\n","            message = entry.children[1]\n","            # Filter out lines with error messages\n","            if message.value != '':\n","                continue\n","            # Filter out empty lines\n","            if line.children[1].value == '':\n","                continue\n","            if line.children[3].value == '':\n","                continue\n","\n","            mod = {'chain': line.children[0].value,\n","                   'index': line.children[1].value,\n","                   'ccd': line.children[3].value}\n","            mod_data.append(mod)\n","\n","\n","    def add_bond_entry(self, b):\n","        \"\"\"\n","        Operates when '+' button is pressed, this time on bond constraints\n","        \"\"\"\n","        # Input for atom1\n","        select_chain1 = widgets.Dropdown(\n","            options=[s['chain'] for s in seq_data],\n","            description='Atom1 chain:',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='150px'))\n","\n","        enter_res1 = widgets.Text(\n","            description='resi: ',\n","            placeholder = '> 0',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='150px'))\n","\n","        show_res1 = widgets.Label(value=\" - \", layout=widgets.Layout(width='100px'))\n","\n","        select_atom1 = widgets.Dropdown(\n","            description='atom: ',\n","            options=[],\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='200px'))\n","\n","        # Input for atom2\n","        select_chain2 = widgets.Dropdown(\n","            options=[s['chain'] for s in seq_data],\n","            description='Atom2 chain:',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='150px'))\n","\n","        enter_res2 = widgets.Text(\n","            description='resi: ',\n","            placeholder = '> 0',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='100px'))\n","\n","        show_res2 = widgets.Label(value=\" - \", layout=widgets.Layout(width='100px'))\n","\n","        select_atom2 = widgets.Dropdown(\n","            description='atom: ',\n","            options=[],\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='200px'))\n","\n","        # Minus button\n","        remove_btn = widgets.Button(description='-', layout=widgets.Layout(width='30px'))\n","        remove_btn.on_click(self.remove_an_entry)\n","\n","        # Error message\n","        message = widgets.HTML(value='', layout=widgets.Layout(width='600px', padding='5px'))\n","\n","        def validate_string(change):\n","            \"\"\"\n","            Validate the input entered by the user.\n","            If the input is incorrect, show the red error message.\n","            \"\"\"\n","            error_type = ''\n","\n","            if select_chain1.value == select_chain2.value and enter_res1.value == enter_res2.value:\n","                error_type = 'duplicate'\n","            else:\n","                show_res1.value = return_res_name(select_chain1.value, enter_res1.value)\n","                show_res2.value = return_res_name(select_chain2.value, enter_res2.value)\n","\n","                atom_list1 = return_atom_name_list(select_chain1.value, enter_res1.value)\n","                atom_list2 = return_atom_name_list(select_chain2.value, enter_res2.value)\n","\n","                if len(atom_list1) == 0:\n","                    error_type = 'index1'\n","                    select_atom1.options = []\n","                else:\n","                    select_atom1.options = atom_list1\n","\n","                if len(atom_list2) == 0:\n","                    error_type = 'index2'\n","                    select_atom2.options = []\n","                else:\n","                    select_atom2.options = atom_list2\n","\n","            if error_type == '':\n","                message.value = \"\"\n","            elif error_type == \"duplicate\":\n","                message.value = f\"<span style='color: red;'>Duplicate residue index.</span>\"\n","            else:\n","                message.value = f\"<span style='color: red;'>Enter the valid number for {error_type}.</span>\"\n","\n","        select_chain1.observe(validate_string, names='value')\n","        enter_res1.observe(validate_string, names='value')\n","        select_atom1.observe(validate_string, names='value')\n","        select_chain2.observe(validate_string, names='value')\n","        enter_res2.observe(validate_string, names='value')\n","        select_atom2.observe(validate_string, names='value')\n","\n","        line = widgets.HBox([select_chain1, enter_res1, show_res1, select_atom1, select_chain2, enter_res2, show_res2, select_atom2, remove_btn])\n","        entry = widgets.VBox([line, message])\n","\n","        self.container.children = list(self.container.children[:-1]) + [entry, self.container.children[-1]]\n","\n","\n","    def update_bond_data(self, b):\n","        \"\"\"\n","        Operates when the confirm button is pressed, this time on bond constraints\n","        \"\"\"\n","        bond_data.clear()\n","        for i in range(1, len(self.container.children)-1):\n","            entry = self.container.children[i]\n","            line = entry.children[0]\n","            message = entry.children[1]\n","            # Filter out lines with error\n","            if message.value != '':\n","                continue\n","            # Filter out empty lines\n","            if line.children[1].value == '':\n","                continue\n","            if line.children[5].value == '':\n","                continue\n","\n","            bond = {'chain1': line.children[0].value,\n","                    'index1': line.children[1].value,\n","                    'atom1': line.children[3].value,\n","                    'chain2': line.children[4].value,\n","                    'index2': line.children[5].value,\n","                    'atom2': line.children[7].value}\n","            bond_data.append(bond)\n","\n","\n","    def add_pocket_entry(self, b):\n","        \"\"\"\n","        Operates when '+' button is pressed, this time on pocket constraints\n","        \"\"\"\n","        receptor = []\n","        for s in seq_data:\n","            if s['chain'] != binder:\n","                receptor.append(s['chain'])\n","\n","        # Reads receptor chain choice\n","        select_chain = widgets.Dropdown(\n","            options=receptor,\n","            description='receptor chain:',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='150px'))\n","\n","        enter_token = None\n","\n","        # print chosen residue name\n","        show_res = widgets.Label(value=\"residue : -\", layout=widgets.Layout(width='200px'))\n","\n","        # Reads max distance in Angstrom\n","        enter_max_distance = widgets.Text(\n","            description='max distance (A) : ',\n","            value='4.0',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='200px'))\n","\n","        # Minus button\n","        remove_btn = widgets.Button(description='-', layout=widgets.Layout(width='30px'))\n","        remove_btn.on_click(self.remove_an_entry)\n","\n","        # Error message\n","        message = widgets.HTML(value='', layout=widgets.Layout(width='600px', padding='5px'))\n","\n","        def update_token_widget(chain_value):\n","            \"\"\"\n","            Update the token widget based on the type of selected chain.\n","            \"\"\"\n","            nonlocal enter_token # Use nonlocal to modify the outer scope variable\n","\n","            if is_it_polymer(chain_value):\n","                enter_token = widgets.Text(\n","                    description='pocket residue index : ',\n","                    placeholder = 'resi > 0',\n","                    style={'description_width': 'initial'},\n","                    layout=widgets.Layout(width='200px'))\n","\n","            else:\n","                # For non-polymers, use atom names from the CCD\n","                ccd_code = None\n","                for s in seq_data:\n","                    if s['chain'] == chain_value and s['type'] == 'ccd':\n","                        ccd_code = s['sequence']\n","                        break\n","                atom_options = []\n","\n","                if ccd_code:\n","                    atom_options = get_atom_names(ccd_code)\n","\n","                enter_token = widgets.Dropdown(\n","                    description='pocket atom name : ',\n","                    options=atom_options,\n","                    style={'description_width': 'initial'},\n","                    layout=widgets.Layout(width='200px'))\n","\n","            enter_token.observe(validate_string, names='value') # Observe the new widget\n","\n","\n","        def validate_string(change):\n","            \"\"\"\n","            Validate the input entered by the user.\n","            If the input is incorrect, show the red error message.\n","            \"\"\"\n","            error_type = ''\n","            chain_value = select_chain.value\n","\n","            if is_it_polymer(chain_value):\n","                # Validation for polymer (residue index)\n","                try:\n","                    index_value = int(enter_token.value)\n","                    if index_value <= 0:\n","                        error_type = 'residue index'\n","                    else:\n","                        for s in seq_data:\n","                            if s['chain'] == chain_value:\n","                                res = get_residue_ccd(s['type'], s['sequence'], enter_token.value)\n","                                if res == '':\n","                                    error_type = 'residue index'\n","                                    show_res.value = 'residue : -'\n","                                else:\n","                                    show_res.value = f\"residue : {res}\"\n","                                break\n","                except ValueError:\n","                    error_type = 'residue index'\n","\n","            else:\n","                # Validation for non-polymer (atom name)\n","                for s in seq_data:\n","                    if s['chain'] == chain_value:\n","                        if s['type'] == 'ccd':\n","                            show_res.value = f\"name: {s['sequence']}\"\n","                            break\n","\n","            for p in pocket_data:\n","                if p['chain'] == chain_value:\n","                    if p['token'] == enter_token.value:\n","                        error_type = \"duplicate\"\n","                        break\n","\n","            try:\n","                max_d = float(enter_max_distance.value)\n","                if max_d <= 0:\n","                    error_type = 'max distance'\n","            except ValueError:\n","                error_type = 'max distance'\n","\n","\n","            if error_type == '':\n","                message.value = \"\"\n","            elif error_type == \"duplicate\":\n","                if is_it_polymer(chain_value):\n","                    message.value = f\"<span style='color: red;'>Duplicate residue index found.</span>\"\n","                else:\n","                    message.value = f\"<span style='color: red;'>Duplicate atom name found.</span>\"\n","            elif error_type == 'max distance':\n","                message.value = f\"<span style='color: red;'>Enter the valid {error_type} value in Angstroms.</span>\"\n","            else:\n","                if is_it_polymer(chain_value):\n","                    message.value = f\"<span style='color: red;'>Enter the valid {error_type}.</span>\"\n","                else:\n","                    message.value = f\"<span style='color: red;'>Enter the valid {error_type} (select from dropdown).</span>\"\n","\n","        # Initialize the enter_token widget based on the initial value of select_chain\n","        update_token_widget(select_chain.value)\n","\n","        select_chain.observe(lambda change: update_token_widget(change['new']), names='value')\n","        enter_max_distance.observe(validate_string, names='value')\n","\n","        line = widgets.HBox([select_chain, enter_token, show_res, enter_max_distance, remove_btn]) # Use enter_token here\n","        entry = widgets.VBox([line, message])\n","\n","        self.container.children = list(self.container.children[:-1]) + [entry, self.container.children[-1]]\n","\n","\n","    def update_pocket_data(self, b):\n","        \"\"\"\n","        Operates when the confirm button is pressed, this time on pocket constraints\n","        \"\"\"\n","        pocket_data.clear()\n","        for i in range(1, len(self.container.children)-1):\n","            entry = self.container.children[i]\n","            line = entry.children[0]\n","            message = entry.children[1]\n","            if message.value != '':\n","                continue\n","            if line.children[1].value == '':\n","                continue # Filter empty text input\n","\n","            pocket = {'chain': line.children[0].value,\n","                      'token': line.children[1].value,\n","                      'max_d': line.children[3].value}\n","            pocket_data.append(pocket)\n","\n","\n","    def add_contact_entry(self, b):\n","        \"\"\"\n","        Operates when '+' button is pressed, this time on contact constraints\n","        \"\"\"\n","        # Input for token1\n","        select_chain1 = widgets.Dropdown(\n","            options=[s['chain'] for s in seq_data],\n","            description='chain1 :',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='150px'))\n","\n","        enter_token1 = None\n","\n","        # print chosen residue name\n","        show_res1 = widgets.Label(value=\"res1: -\", layout=widgets.Layout(width='200px'))\n","\n","        # Input for token2\n","        select_chain2 = widgets.Dropdown(\n","            options=[s['chain'] for s in seq_data],\n","            description='chain2 :',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='150px'))\n","\n","        enter_token2 = None\n","\n","        # print chosen residue name\n","        show_res2 = widgets.Label(value=\"res2: -\", layout=widgets.Layout(width='200px'))\n","\n","        # Reads max distance in Angstrom\n","        enter_max_distance = widgets.Text(\n","            description='max distance (A) : ',\n","            value='4.0', # defaults to 4.0 A\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='200px'))\n","\n","        # Minus button\n","        remove_btn = widgets.Button(description='-', layout=widgets.Layout(width='30px'))\n","        remove_btn.on_click(self.remove_an_entry)\n","\n","        # Error message\n","        message = widgets.HTML(value='', layout=widgets.Layout(width='600px', padding='5px'))\n","\n","        def update_token_widgets(chain_values):\n","            \"\"\"\n","            Update the token widget based on the type of selected chain.\n","            \"\"\"\n","            nonlocal enter_token1, enter_token2\n","\n","            if is_it_polymer(chain_values[0]):\n","                enter_token1 = widgets.Text(\n","                    description='resi1: ',\n","                    placeholder = '> 0',\n","                    style={'description_width': 'initial'},\n","                    layout=widgets.Layout(width='200px'))\n","            else:\n","                ccd_code = None\n","                for s in seq_data:\n","                    if s['chain'] == chain_values[0] and s['type'] == 'ccd':\n","                        ccd_code = s['sequence']\n","                        break\n","                if ccd_code:\n","                    enter_token1 = widgets.Dropdown(\n","                        description='atom1: ',\n","                        options=get_atom_names(ccd_code),\n","                        style={'description_width': 'initial'},\n","                        layout=widgets.Layout(width='200px'))\n","\n","            if is_it_polymer(chain_values[1]):\n","                enter_token2 = widgets.Text(\n","                    description='resi2: ',\n","                    placeholder = '> 0',\n","                    style={'description_width': 'initial'},\n","                    layout=widgets.Layout(width='200px'))\n","            else:\n","                ccd_code = None\n","                for s in seq_data:\n","                    if s['chain'] == chain_values[1] and s['type'] == 'ccd':\n","                        ccd_code = s['sequence']\n","                        break\n","                if ccd_code:\n","                    enter_token2 = widgets.Dropdown(\n","                        description='atom2: ',\n","                        options=get_atom_names(ccd_code),\n","                        style={'description_width': 'initial'},\n","                        layout=widgets.Layout(width='200px'))\n","\n","            enter_token1.observe(validate_string, names='value')\n","            enter_token2.observe(validate_string, names='value')\n","\n","\n","        def validate_string(change):\n","            \"\"\"\n","            Validate the input entered by the user.\n","            If the input is incorrect, show the red error message.\n","            \"\"\"\n","            error_type = ''\n","            chain_values = [select_chain1.value, select_chain2.value]\n","\n","            if select_chain1.value == select_chain2.value and enter_token1.value == enter_token2.value:\n","                error_type = 'duplicate'\n","\n","            else:\n","                for i in range(2):\n","                    chain_value = chain_values[i]\n","                    token_widget = enter_token1 if i == 0 else enter_token2\n","\n","                    if is_it_polymer(chain_value):\n","                        # Validation for polymer (residue index)\n","                        res = return_res_name(chain_value, token_widget.value)\n","                        if res == '':\n","                            error_type = f'residue{i+1} index'\n","                            show_res_value = f'res{i+1}: -'\n","                        else:\n","                            show_res_value = f\"res{i+1}: {res}\"\n","\n","                    else:\n","                        # Validation for non-polymer (atom name)\n","                        for s in seq_data:\n","                            if s['chain'] == chain_value:\n","                                if s['type'] == 'ccd':\n","                                    show_res_value = f\"name: {s['sequence']}\"\n","                                    break\n","                                else:\n","                                    show_res_value = f\"res{i+1}: -\"\n","                                    break\n","                    if i == 0:\n","                        show_res1.value = show_res_value\n","                    else:\n","                        show_res2.value = show_res_value\n","\n","            for c in contact_data:\n","                if c['chain1'] == chain_values[0] and c['chain2'] == chain_values[1]:\n","                    if c['token1'] == enter_token1.value and c['token2'] == enter_token2.value:\n","                        error_type = \"duplicate\"\n","                        break\n","\n","            try:\n","                max_d = float(enter_max_distance.value)\n","                if max_d <= 0:\n","                    error_type = 'max distance'\n","            except ValueError:\n","                error_type = 'max distance'\n","\n","            if error_type == '':\n","                message.value = \"\"\n","            elif error_type == \"duplicate\":\n","                message.value = f\"<span style='color: red;'>Duplicate info entered.</span>\"\n","            else:\n","                message.value = f\"<span style='color: red;'>Enter the valid number for {error_type}.</span>\"\n","\n","        update_token_widgets([select_chain1.value, select_chain2.value])\n","\n","        select_chain1.observe(lambda change: update_token_widgets([change['new'], select_chain2.value]), names='value')\n","        select_chain2.observe(lambda change: update_token_widgets([select_chain1.value, change['new']]), names='value')\n","\n","        enter_token1.observe(validate_string, names='value')\n","        enter_token2.observe(validate_string, names='value')\n","        enter_max_distance.observe(validate_string, names='value')\n","\n","        line = widgets.HBox([select_chain1, enter_token1, show_res1, select_chain2, enter_token2, show_res2, enter_max_distance, remove_btn])\n","        entry = widgets.VBox([line, message])\n","\n","        self.container.children = list(self.container.children[:-1]) + [entry, self.container.children[-1]]\n","\n","\n","    def update_contact_data(self, b):\n","        \"\"\"\n","        Operates when the confirm button is pressed, this time on pocket constraints\n","        \"\"\"\n","        contact_data.clear()\n","        for i in range(1, len(self.container.children)-1):\n","            entry = self.container.children[i]\n","            line = entry.children[0]\n","            message = entry.children[1]\n","            if message.value != '':\n","                continue\n","            if line.children[1].value == '':\n","                continue\n","            if line.children[4].value == '':\n","                continue\n","\n","            contact = {'chain1': line.children[0].value,\n","                       'token1': line.children[1].value,\n","                       'chain2' : line.children[3].value,\n","                       'token2' : line.children[4].value,\n","                       'max_d': line.children[6].value}\n","            contact_data.append(contact)\n","\n","\n","    def add_template_entry(self, b):\n","        \"\"\"\n","        Operates when '+' button is pressed, this time on templates\n","        \"\"\"\n","        # template file\n","        template_file = widgets.Dropdown(\n","            options=[t['cif'] for t in template_data],\n","            description='File name :',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='300px'))\n","\n","        protein_chain = widgets.Dropdown(\n","            options=[s['chain'] for s in seq_data if s['type'] == 'protein'],\n","            description='Protein chain ID :',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='200px'))\n","\n","        template_chain = widgets.Dropdown(\n","            options=['Auto'],\n","            description='Template chain ID :',\n","            style={'description_width': 'initial'},\n","            layout=widgets.Layout(width='200px'))\n","\n","        # minus button\n","        remove_btn = widgets.Button(description='-', layout=widgets.Layout(width='30px'))\n","        remove_btn.on_click(self.remove_seq_entry)\n","\n","\n","        def update_chains(change):\n","            template_chain.options = ['Auto']+get_polypeptide_chain_ids(template_file.value)\n","\n","        template_file.observe(update_chains, names='value')\n","\n","        line = widgets.HBox([template_file, protein_chain, template_chain, remove_btn])\n","        entry = widgets.VBox([line])\n","\n","        self.container.children = list(self.container.children[:-1]) + [entry, self.container.children[-1]]\n","\n","\n","    def update_template_data(self, b):\n","        \"\"\"\n","        Operates when the confirm button is pressed\n","        \"\"\"\n","        for i in range(1, len(self.container.children)-1):\n","            line = self.container.children[i]\n","            for t in template_data:\n","                if t['cif'] == line.children[0].value:\n","                    t['protein'] = line.children[1].value\n","                    if line.children[2].value != 'Auto':\n","                        t['template'] = line.children[2].value\n","                    break\n","\n","\n","title = widgets.HTML(\"<h4>Click the plus button to add molecules, and minus button to remove ones. Click the confirm button after entering all entries.</h4>\")\n","add_button = widgets.Button(description='+', layout=widgets.Layout(width='30px'))\n","confirm_button = widgets.Button(description='confirm', layout=widgets.Layout(width='100px'))\n","buttons = widgets.HBox([add_button, confirm_button])\n","seq_container = widgets.VBox([title, buttons])\n","\n","add_new_seq = modify_entries(seq_container)\n","\n","add_button.on_click(add_new_seq.add_seq_entry)\n","confirm_button.on_click(add_new_seq.update_seq_data)\n","display(seq_container)"],"metadata":{"id":"5574tPYsUZxo"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## (Optional) Enter modification / constraint data"],"metadata":{"id":"_2ro0HncbFfL"}},{"cell_type":"code","source":["#@title Residue-wise Modifications\n","#@markdown Select the chain, and type the 1-based index for the residue to modify. Type the CCD code for the modified residue.\n","\n","assert len(seq_data) > 0, \"No molecule info entered\"\n","pol_flag = False\n","for s in seq_data:\n","    if s['type'] in ['protein', 'dna', 'rna']:\n","        pol_flag = True\n","        break\n","assert pol_flag, \"No polymer molecules were entered\"\n","\n","title = widgets.HTML(\"<h4>Click the plus button to modify biopolymers you entered above, and minus button to remove modifications. Click the confirm button after entering all entries.</h4>\")\n","add_button = widgets.Button(description='+', layout=widgets.Layout(width='30px'))\n","confirm_button = widgets.Button(description='confirm', layout=widgets.Layout(width='100px'))\n","buttons = widgets.HBox([add_button, confirm_button])\n","mod_container = widgets.VBox([title, buttons])\n","\n","add_new_mod = modify_entries(mod_container)\n","\n","add_button.on_click(add_new_mod.add_mod_entry)\n","confirm_button.on_click(add_new_mod.update_mod_data)\n","display(mod_container)"],"metadata":{"id":"yNkc63uHx0VM","colab":{"base_uri":"https://localhost:8080/","height":211},"outputId":"46129a76-c79e-42a8-97b9-d0318896bcda","cellView":"form","collapsed":true,"executionInfo":{"status":"error","timestamp":1765827192572,"user_tz":300,"elapsed":14,"user":{"displayName":"Joe Z Wu","userId":"04248511503294646969"}}},"execution_count":null,"outputs":[{"output_type":"error","ename":"AssertionError","evalue":"No molecule info entered","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)","\u001b[0;32m/tmp/ipython-input-3352276732.py\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m#@markdown Select the chain, and type the 1-based index for the residue to modify. Type the CCD code for the modified residue.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0;32massert\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq_data\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"No molecule info entered\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mpol_flag\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0ms\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mseq_data\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mAssertionError\u001b[0m: No molecule info entered"]}]},{"cell_type":"code","source":["#@title Bond Constraints\n","#@markdown Only bonds between two atoms from biopolymers can be entered. \\\n","#@markdown Select the the atom's chain ID, enter 1-based residue index, and then select the atom from the dropdown menu.\n","\n","assert len(seq_data) > 0, \"No molecule info entered\"\n","\n","title = widgets.HTML(\"<h4>Click the plus button to add covalent bond constraints to the molecules, and minus button to remove constraints. Click the confirm button after entering all entries.</h4>\")\n","add_button = widgets.Button(description='+', layout=widgets.Layout(width='30px'))\n","confirm_button = widgets.Button(description='confirm', layout=widgets.Layout(width='100px'))\n","buttons = widgets.HBox([add_button, confirm_button])\n","bond_container = widgets.VBox([title, buttons])\n","\n","add_new_bond = modify_entries(bond_container)\n","\n","add_button.on_click(add_new_bond.add_bond_entry)\n","confirm_button.on_click(add_new_bond.update_bond_data)\n","display(bond_container)"],"metadata":{"id":"eVlee2I1Gziy","cellView":"form"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@title Pocket Constraints\n","#@markdown First select the chain ID of the molecule of interest in the pocket (binder). \\\n","#@markdown Then enter the info for the residue/atoms that constitute the pocket. \\\n","#@markdown You can also set the maximum distance between the binder and the pocket in Angstroms. \\\n","#@markdown Note that only one binder per system is supported.\n","\n","assert len(seq_data) > 1, \"Less than 2 molecules entered\"\n","\n","title = widgets.HTML(\"<h4>Choose the binder chain. Then add the pocket constraints. Click the confirm button after adding all entries.</h4>\")\n","choose_ligand = widgets.Dropdown(\n","    options=[s['chain'] for s in seq_data],\n","    description='binder chain ID:',\n","    style={'description_width': 'initial'},\n","    layout=widgets.Layout(width='150px'))\n","add_button = widgets.Button(description='+', layout=widgets.Layout(width='30px'))\n","confirm_button = widgets.Button(description='confirm', layout=widgets.Layout(width='100px'))\n","header = widgets.VBox([title, choose_ligand])\n","buttons = widgets.HBox([add_button, confirm_button])\n","pocket_container = widgets.VBox([header, buttons])\n","\n","def update_binder(change):\n","    global binder\n","    binder = change['new']\n","\n","choose_ligand.observe(update_binder, names='value')\n","\n","add_new_pocket = modify_entries(pocket_container)\n","add_button.on_click(add_new_pocket.add_pocket_entry)\n","confirm_button.on_click(add_new_pocket.update_pocket_data)\n","display(pocket_container)"],"metadata":{"cellView":"form","id":"2UOMAzHpEswe","collapsed":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@title Contact Constraints\n","#@markdown Select a pair of residue / atom and set the maximum distance between them.\n","\n","assert len(seq_data) > 0, \"No molecule info entered\"\n","\n","title = widgets.HTML(\"<h4>Click the plus button to add contact constraints to the molecules, and minus button to remove constraints. Click the confirm button after entering all entries.</h4>\")\n","add_button = widgets.Button(description='+', layout=widgets.Layout(width='30px'))\n","confirm_button = widgets.Button(description='confirm', layout=widgets.Layout(width='100px'))\n","buttons = widgets.HBox([add_button, confirm_button])\n","contact_container = widgets.VBox([title, buttons])\n","\n","add_new_contact = modify_entries(contact_container)\n","add_button.on_click(add_new_contact.add_contact_entry)\n","confirm_button.on_click(add_new_contact.update_contact_data)\n","display(contact_container)"],"metadata":{"cellView":"form","id":"UAr0-Q8kfbNT","collapsed":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## (Optional) Upload template / custom MSA files for proteins"],"metadata":{"id":"mhGnG1GEglwB"}},{"cell_type":"code","source":["#@markdown Run this cell to upload template cif files. You can select multiple files.\n","\n","assert seq_data, \"No molecule info entered\"\n","\n","proteins = [s for s in seq_data if s['type'] == 'protein']\n","assert proteins, \"No protein molecules entered! Templates are currently only applicable to proteins.\"\n","\n","\n","cif_files = files.upload()\n","template_data.clear()\n","\n","for c in list(cif_files.keys()):\n","    chainIDs = get_polypeptide_chain_ids(c)\n","    if len(chainIDs) > 0:\n","        template_data.append({'cif':c, 'protein':None, 'template':None})\n","\n","assert template_data, \"No polypeptide chains were found in the uploaded cif files.\"\n","\n","title = widgets.HTML(\"<h4>Uploaded template files are automatically matched in default. To manually set the chain ID click the plus button below.</h4>\")\n","add_button = widgets.Button(description='+', layout=widgets.Layout(width='30px'))\n","confirm_button = widgets.Button(description='confirm', layout=widgets.Layout(width='100px'))\n","buttons = widgets.HBox([add_button, confirm_button])\n","template_container = widgets.VBox([title, buttons])\n","\n","add_new_template = modify_entries(template_container)\n","\n","add_button.on_click(add_new_template.add_template_entry)\n","confirm_button.on_click(add_new_template.update_template_data)\n","display(template_container)"],"metadata":{"cellView":"form","id":"gB09x71GhKdO","collapsed":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@markdown Upload custom MSA a3m files and run this cell to allocate them.\n","\n","assert len(seq_data) > 0, \"No molecule info entered\"\n","\n","proteins = [s for s in seq_data if s['type'] == 'protein']\n","assert len(proteins) > 0, \"No protein molecules entered! MSAs are currently only applicable to proteins.\"\n","\n","title = widgets.HTML(\"<h4>Upload your MSA files and enter the file name next to the corresponding polypeptide chain.</h4>\")\n","\n","p_chains = []\n","for p in proteins:\n","    cyclic = 'cyclic' if p['cyclic'] else 'linear'\n","    chain = widgets.Label(value=f\"chain ID : {p['chain']}  length : {len(p['sequence'])} aa, {cyclic}\", layout=widgets.Layout(width='500px'))\n","    filename = widgets.Text(description='MSA file name : ', value='.a3m', style={'description_width': 'initial'}, layout=widgets.Layout(width='300px'))\n","    p_chains.append(widgets.HBox([chain, filename]))\n","\n","table = widgets.VBox(p_chains)\n","\n","confirm_button = widgets.Button(description='confirm', layout=widgets.Layout(width='100px'))\n","\n","def update_msa_data(b):\n","    for i, p in enumerate(proteins):\n","        fname = table.children[i].children[1].value.split('.')[0]\n","        if fname == '':\n","            continue\n","        else:\n","            p['msa'] = f\"/content/{fname}.a3m\"\n","\n","confirm_button.on_click(update_msa_data)\n","display(title, table, confirm_button)"],"metadata":{"id":"WOQ9ldNqnbEa"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## (Optional) Predict binding affinity"],"metadata":{"id":"EjpvgFJEPdlB"}},{"cell_type":"code","source":["assert seq_data, \"No molecule info entered\"\n","\n","ligands = [s for s in seq_data if s['type'] in ['smiles', 'ccd']]\n","assert ligands, \"No small molecules entered! Binding affinity predictions cannot be applied to proteins and nucleic acids.\"\n","\n","import io\n","\n","#@title Set the target ligand for affinity prediction\n","\n","#@markdown Number of diffusion samples to be used for affinity prediction. (default: 5)\n","diffusion_samples_affinity = 5 #@param {type:\"slider\", min:1, max:10, step:1}\n","#@markdown Number of sampling steps for affinity prediction. (default: 200)\n","sampling_steps_affinity = 200 #@param {type:\"slider\", min:50, max:400, step:50}\n","\n","lig_select = widgets.Dropdown(options=['None']+[l['chain'] for l in ligands],\n","                              value='None',\n","                              style={'description_width': 'initial'},\n","                              description='Ligand chain ID:',\n","                              layout=widgets.Layout(width='300px'))\n","\n","lig_image = widgets.Image(value=b'', format='png', width=400, height=300)\n","\n","def update_sdf(change): # callback function to interactively update viewer\n","    ligand = lig_select.value\n","    if ligand == 'None':\n","        return\n","    for s in seq_data:\n","        if s['chain'] == ligand:\n","            RDLogger.DisableLog('rdApp.*')\n","            if s['type'] == 'smiles':\n","                mol = Chem.MolFromSmiles(s['sequence'], sanitize=True)\n","            elif s['type'] == 'ccd':\n","                ccd = s['sequence'].upper()\n","                if not os.path.exists(ccd+'_ideal.sdf'):\n","                    subprocess.run(f'wget https://files.rcsb.org/ligands/download/{ccd}_ideal.sdf', shell=True)\n","                sdf_file = f'{ccd}_ideal.sdf'\n","                mol = Chem.rdmolfiles.SDMolSupplier(sdf_file)[0]\n","            break\n","\n","    AllChem.Compute2DCoords(mol)\n","    img = Draw.MolToImage(mol, size=(400, 300))\n","\n","    with io.BytesIO() as output:\n","        img.save(output, format=\"PNG\")\n","        lig_image.value = output.getvalue()\n","\n","lig_select.observe(update_sdf, names='value')\n","\n","print(\"Select the ligand's chain ID and check the structure. Select None to cancel.\")\n","\n","display(lig_select, lig_image)"],"metadata":{"cellView":"form","id":"jpoXlc-ZPoMd"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":[],"metadata":{"id":"3ELJhfjoUVj9"}},{"cell_type":"markdown","source":["# Run Prediction"],"metadata":{"id":"G6XxZePbbV93"}},{"cell_type":"code","source":["#@title Create YAML file from the input data\n","#@markdown Once you are confident with all the inputs above, run this cell to generate the input file.\n","\n","#@markdown The created .yaml file can be downloaded, edited and uploaded for faster setup in the future.\n","\n","import yaml\n","\n","data = {'version': 1, 'sequences':[]}\n","\n","for s in seq_data:\n","    if s['type'] in ['protein', 'dna', 'rna']:\n","        seq = {s['type']:{'id':s['chain'], 'sequence':s['sequence']}}\n","        mod = []\n","        for m in mod_data:\n","            if m['chain'] == s['chain']:\n","                mod.append({'position':int(m['index']), 'ccd':m['ccd']})\n","        if mod:\n","            seq[s['type']]['modifications'] = mod\n","        if s['type'] == 'protein' and s['msa'] != '':\n","            seq['protein']['msa'] = s['msa']\n","        seq[s['type']]['cyclic'] = s['cyclic']\n","    elif s['type'] == 'smiles':\n","        seq = {'ligand':{'id':s['chain'], 'smiles':s['sequence']}}\n","    elif s['type'] == 'ccd':\n","        seq = {'ligand':{'id':s['chain'], 'ccd':s['sequence']}}\n","    data['sequences'].append(seq)\n","\n","if (len(bond_data)+len(pocket_data)+len(contact_data)) >0:\n","    data['constraints'] = []\n","\n","    for bond in bond_data:\n","        data['constraints'].append({'bond':{'atom1':[bond['chain1'], int(bond['index1']), bond['atom1']], 'atom2':[bond['chain2'], int(bond['index2']), bond['atom2']]}})\n","\n","    if len(pocket_data) > 0:\n","        data['constraints'].append({'pocket':{'binder':binder, 'contacts':[], 'max_distance':''}})\n","\n","    max_distance = 0.0\n","    for pocket in pocket_data:\n","        if is_it_polymer(pocket['chain']):\n","            token = int(pocket['token'])\n","        else:\n","            token = pocket['token']\n","        data['constraints'][-1]['pocket']['contacts'].append([pocket['chain'], token])\n","        max_d = float(pocket['max_d'])\n","\n","        if max_d > max_distance:\n","            max_distance = max_d\n","\n","    data['constraints'][-1]['pocket']['max_distance'] = max_distance\n","\n","    for contact in contact_data:\n","        if is_it_polymer(contact['chain1']):\n","            token1 = int(contact['token1'])\n","        else:\n","            token1 = contact['token1']\n","        if is_it_polymer(contact['chain2']):\n","            token2 = int(contact['token2'])\n","        else:\n","            token2 = contact['token2']\n","\n","        data['constraints'].append({'contact':{'token1':[contact['chain1'], token1], 'token2':[contact['chain2'], token2], 'max_distance':float(contact['max_d'])}})\n","\n","if len(template_data) >0:\n","    data['templates'] = []\n","    for t in template_data:\n","        temp = {'cif':'/content/'+t['cif']}\n","        if t['protein'] is not None:\n","            temp['chain_id'] = t['protein']\n","        if t['template'] is not None and t['template'] != 'Auto':\n","            temp['template_id'] = t['template']\n","        data['templates'].append(temp)\n","\n","if lig_select is not None and lig_select.value != 'None':\n","    data['properties'] = [{'affinity':{'binder':lig_select.value}}]\n","\n","with open(f'{job_title}.yaml', 'w') as f:\n","    yaml.dump(data, f, default_flow_style=False, sort_keys=False)\n","    print('Done!')"],"metadata":{"cellView":"form","id":"7hBARuPAIFOn","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1765836938828,"user_tz":300,"elapsed":52,"user":{"displayName":"Joe Z Wu","userId":"04248511503294646969"}},"outputId":"291d1991-5e56-46ab-9891-be5246887be7"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Done!\n"]}]},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive')"],"metadata":{"id":"k6IRx0tgVkfG"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["import os\n","os.environ['TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD'] = '1'\n","#@title Run prediction using Boltz-2\n","#@markdown Output format\n","output_format = 'mmcif' #@param [\"pdb\", \"mmcif\"]\n","#@markdown Number of dataloader workers\n","num_workers = 2 #@param {type:\"slider\", min:0, max:5, step:1}\n","#@markdown Lower the step scale to increase the diversity of result. (default: 1.638)\n","step_scale = 1.638 #@param {type:\"slider\", min:1, max:2, step:0.001}\n","#@markdown Number of diffusion samples to be generated. (default: 1, AlphaFold3: 5)\n","diffusion_samples = 4 #@param {type:\"slider\", min:1, max:10, step:1}\n","#@markdown Number of recycling steps for the prediction. (default: 3, AlphaFold3: 10)\n","recycling_steps = 3 #@param {type:\"slider\", min:1, max:25, step:1}\n","#@markdown Number of sampling steps for structure prediction. (default: 200)\n","sampling_steps = 50 #@param {type:\"slider\", min:50, max:400, step:50}\n","#@markdown Maximum number of MSA sequences to be used\n","max_msa_seqs = 8192 #@param [32, 64, 128, 256, 512, 1024, 2048, 4096, 8192]\n","#@markdown Subsample MSA?\n","subsample_msa = False #@param {type:\"boolean\"}\n","#@markdown Number of subsampled MSA\n","num_subsampled_msa = 1024 #@param [4, 8, 16, 32, 64, 128, 256, 512, 1024]\n","#@markdown MSA pairing strategy\n","msa_pairing_strategy = 'greedy' #@param ['greedy', 'complete']\n","\n","use_msa_server = False\n","for s in seq_data:\n","    if s['type'] == 'protein' and s['msa'] == '':\n","        use_msa_server = True\n","        break\n","\n","\n","\n","commandline = f'{job_title}.yaml --num_workers {num_workers} --step_scale {step_scale} --recycling_steps {recycling_steps} --sampling_steps {sampling_steps}'\n","commandline += f' --diffusion_samples {diffusion_samples} --max_msa_seqs {max_msa_seqs} --msa_pairing_strategy {msa_pairing_strategy}'\n","if subsample_msa:\n","    commandline += f' --num_subsampled_msa {num_subsampled_msa}'\n","if use_msa_server:\n","    commandline += ' --use_msa_server'\n","if lig_select is not None and lig_select.value != 'None':\n","    commandline += f' --diffusion_samples_affinity {diffusion_samples_affinity} --sampling_steps_affinity {sampling_steps_affinity}'\n","if runtime.startswith('GPU'):\n","    commandline += ' --no_kernels --accelerator gpu'\n","elif runtime == 'TPU':\n","    commandline += ' --accelerator tpu'\n","else:\n","    commandline += ' --accelerator cpu'\n","\n","!boltz predict {commandline} --out_dir /content/{job_title}"],"metadata":{"id":"luRtz_wunbMW","cellView":"form"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@title Download results as a zip file\n","\n","import zipfile\n","from google.colab import files\n","\n","print(\"Downloading result files.\")\n","filename = f'{job_title}.zip'\n","\n","with zipfile.ZipFile(filename, 'w') as zip_file:\n","    dir_path = f'/content/{job_title}/boltz_results_{job_title}'\n","    for root, directory, items in os.walk(dir_path):\n","        for item in items:\n","            path = os.path.join(root, item)\n","            zip_file.write(path, arcname=os.path.relpath(os.path.join(root, item), dir_path), compress_type=zipfile.ZIP_DEFLATED)\n","\n","files.download(filename)"],"metadata":{"id":"Kay_lp4EQlok","cellView":"form","colab":{"base_uri":"https://localhost:8080/","height":34},"executionInfo":{"status":"ok","timestamp":1765838612847,"user_tz":300,"elapsed":751,"user":{"displayName":"Joe Z Wu","userId":"04248511503294646969"}},"outputId":"b156985b-1332-4352-ebe6-0a9bf7ab426e"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Downloading result files.\n"]},{"output_type":"display_data","data":{"text/plain":["<IPython.core.display.Javascript object>"],"application/javascript":["\n","    async function download(id, filename, size) {\n","      if (!google.colab.kernel.accessAllowed) {\n","        return;\n","      }\n","      const div = document.createElement('div');\n","      const label = document.createElement('label');\n","      label.textContent = `Downloading \"${filename}\": `;\n","      div.appendChild(label);\n","      const progress = document.createElement('progress');\n","      progress.max = size;\n","      div.appendChild(progress);\n","      document.body.appendChild(div);\n","\n","      const buffers = [];\n","      let downloaded = 0;\n","\n","      const channel = await google.colab.kernel.comms.open(id);\n","      // Send a message to notify the kernel that we're ready.\n","      channel.send({})\n","\n","      for await (const message of channel.messages) {\n","        // Send a message to notify the kernel that we're ready.\n","        channel.send({})\n","        if (message.buffers) {\n","          for (const buffer of message.buffers) {\n","            buffers.push(buffer);\n","            downloaded += buffer.byteLength;\n","            progress.value = downloaded;\n","          }\n","        }\n","      }\n","      const blob = new Blob(buffers, {type: 'application/binary'});\n","      const a = document.createElement('a');\n","      a.href = window.URL.createObjectURL(blob);\n","      a.download = filename;\n","      div.appendChild(a);\n","      a.click();\n","      div.remove();\n","    }\n","  "]},"metadata":{}},{"output_type":"display_data","data":{"text/plain":["<IPython.core.display.Javascript object>"],"application/javascript":["download(\"download_aefacdf7-cb9c-4327-907d-4b9e5ff100e5\", \"hpk1-lig1.zip\", 9934356)"]},"metadata":{}}]},{"cell_type":"code","source":["#@title (Optional) Calculate Ligand Kd and dG\n","\n","assert lig_select is not None, \"No ligand selected\"\n","\n","import json\n","import math\n","\n","#@markdown Set the temperature in Celsius.\n","temperature = 25 #@param {type:\"slider\", min:0, max:100, step:1}\n","\n","file_path = f'/content/{job_title}/boltz_results_{job_title}/predictions/{job_title}/affinity_{job_title}.json'\n","\n","with open(file_path, 'r', encoding='utf-8') as f:\n","    raw_data = json.load(f)\n","\n","    pred_value = raw_data['affinity_pred_value']\n","    Kd = math.pow(10, pred_value)/1000000\n","    dG = 8.3144626*(273.15+temperature)*math.log(Kd)/1000\n","\n","print(f'Predicted dissociation eq constant  : {Kd:.3e} mol/L  (= {Kd*1000000:.1f} uM)')\n","print(f'Predicted binding Gibbs free energy : {dG:.3f} kJ/mol')\n","print(f'                                      {dG/4.184:.3f} kcal/mol')"],"metadata":{"cellView":"form","id":"fXLz9TEojkBT"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# Tools used"],"metadata":{"id":"0jgu3H061gPp"}},{"cell_type":"markdown","source":["* Boltz-2\n","> S. Passaro, G. Corso, J. Wohlwend, M. Reveiz, S. Thaler, V. R. Somnath, N. Getz, T. Portnoi, J. Roy, H. Stark, D. Kwabi-Addo, D. Beaini, T. Jaakkola, and R. Barzilay (2025) \"Boltz-2: Towards Accurate and Efficient Binding Affinity Prediction.\" bioRxiv.\n","\n","* Boltz-1\n","> J. Wohlwend, G. Corso, S. Passaro, M. Reveiz, K. Leidal, W. Swiderski,  T. Portnoi, I. Chinn, J. Silterra, T. Jaakkola, and R. Barzilay (2024) \"Boltz-1: Democratizing Biomolecular Interaction Modeling.\" bioRxiv. DOI:[10.1101/2024.11.19.624167](https://doi.org/10.1101/2024.11.19.624167)\n","\n","* ColabFold\n","> M. Mirdita, K. Schtze, Y. Moriwaki, L. Heo, S. Ovchinnikov, and M. Steinegger (2022) \"ColabFold: making protein folding accessible to all\" Nature methods, 19, 679-682. DOI:[10.1038/s41592-022-01488-1](https://doi.org/10.1038/s41592-022-01488-1)\n","\n","*  RDKit\n","> RDKit: Open-source cheminformatics. https://www.rdkit.org\n"],"metadata":{"id":"0b8T8nvcdV8L"}},{"cell_type":"markdown","source":["# 3.2 Structure Preparation in Maestro\n","\n","1. On the laptop, start Maestro in the working directory we used earlier in Step `2.1` and open the project we set up (`hpk1.prj`)\n","2. unzip one of the Boltz prediction zip files in a temporary folder.\n","3. Import the Bolt-predicted structure. Find it under `predictions/JOBNAME/JOBNAME_model_0.cif`, where `JOBNAME` is the job name you picked when predicted the structure above; `hpk1-lig1` for example.\n","4. Compare the Boltz-predicted structure with one of the deposited PDB crystal structures of this system, `7L25`, for example.\n","      *    Download the `7L25` PDB strucure using the `File--Get PDB` feature.\n","      *    Split the structure into chains.\n","      *    Superimpose the protein structure of the Boltz-predicted structure with one of the chains of the PDB structure using the `Superposition--Calpha-atoms` Task. How different are the two structures?\n","      *    Split the structures into protein, ligand, etc.\n","      *    Are the two ligands well aligned? Measure the RMSD deviation between the two ligands using the `Superposition--Maximum Common Substructure` Task.\n","      *    Are the main protein-ligand interactions present in the predicted structure?\n","6. Double-check the bond orders and formal charges of the ligand. Fix any issues.\n","7. Add hydrogen atoms to the receptor structure and the ligand using the `Build--Add Hydrogens` feature.\n","8. Export the protein structure to a `.pdb` file and the ligand to a `.sdf` file.\n","9. Join the protein and ligand structure into a new workspace entry using `Workspace--Create Entry` feature. Export the structure of the complex into a `.pdb` file.\n","10. Repeat the above with the Boltz prediction for the second ligand.\n","\n"],"metadata":{"id":"cgaNL7yTISfO"}}]}